### ISSUE TITLE
The story of safestack.h and lhash.h

### ISSUE THREAD

**Issue by levitte** [openssl/openssl#8102]
`include/openssl/safestack.h` and `include/openssl/lhash.h` contain macros to generate static inline stack and lhash functions tied to specific types, which are simply wrappers around `OPENSSL_sk_` and `OPENSSL_lh_` functions.  This is in itself fine, the macros are designed to be used in code that needs those structures and functions.

`include/openssl/safestack.h` and `include/openssl/lhash.h` also contain uses of those macros to generate the stack and lhash functions around common types.  This is more problematic, because any code including these header files may be forced to be linked with libcrypto, since these macro expansions add references to `OPENSSL_sk_` and `OPENSSL_lh_` functions, even if the code in question only declares a variable a `STACK_OF(FOO)`.  Some compilers are forgiving and remove dead code, masking this issue in the process.

The issue with the uses of the macros in those headers has been solved for SunCC by making the `OPENSSL_sk_` and `OPENSSL_lh_` symbols weak for that compiler.  However, that doesn't resolve the issue in a more general way.

Related to #7721 and #8087, as well as the originating issue #6912

**Comment by blastwave** [comment]
It may be worth while to test this on Solaris x86 and Sparc and also Linux with the Oracle Studio 12.6 compiler tools.

**Comment by makr** [comment]
... and with ```gcc -fkeep-inline-functions -g``` (see #6912).

**Comment by paulidale** [comment]
Suggestion: add an `OPENSSL_NO_HEADER_INLINE` guard around the expansion of the various DEFINE macros in the _header_ files.  Modify all of the OpenSSL source so that this is defined and the DEFINE macro expansions get moved elsewhere.

Existing application will work exactly as they do now, those with problems with inline function can define the guard and the problems are gone.

**Comment by levitte** [comment]
> Modify all of the OpenSSL source so that this is defined

Uhmmm, you lost me there.  Can I assume that you're talking about the few test programs that include safestack.h or lhash.h and do not get linked with libcrypto?

**Comment by paulidale** [comment]
I was talking about _user_ applications that use safestack (it is a public API).
Our applications (tests, app/*) can be fixed to accommodate any changes we make.

I'm suggesting we change the end of safestack.h to:
```
#ifndef OPENSSL_NO_HEADER_INLINE
DEFINE_SPECIAL_STACK_OF(OPENSSL_STRING, char)
DEFINE_SPECIAL_STACK_OF_CONST(OPENSSL_CSTRING, char)
DEFINE_SPECIAL_STACK_OF(OPENSSL_BLOCK, void)
#endif
```
and compile OpenSSL with `-DOPENSSL_NO_HEADER_INLINE` (or equivalent).

A user who includes `safestack.h` _will_ get the definitions of these inline functions, so we stay backward compatible.  If this causes a problem with their compiler, they can define `OPENSSL_NO_HEADER_INLINE` to avoid it.  Since it is a problem, they will need to make code changes anyway, so we're not breaking the API.

There are a number of other public headers that use `DEFINE_STACK_OF` and `DEFINE_LHASH_OF`. 
 These should be addressed at the same time.

**Comment by levitte** [comment]
> and compile OpenSSL with `-DOPENSSL_NO_HEADER_INLINE` (or equivalent).

Why?  *The OpenSSL source doesn't have a problem with those DEFINE expansions*, save a couple of test programs.  I'd suggest we selectively add a `#define OPENSSL_NO_HEADER_INLINE` in the programs that actually need it.

What we should do is have a build with `-fkeep-inline-functions`, for the purpose of catching this type of mistake (I believe it was a mistake to have the DEFINE expansions in safestack.h and lhash.h)

**Comment by blastwave** [comment]
Please don't post 20,000 lines of repetition when we can get the point with just ten lines of examples. I think we should talk about the actual C standard a bit here and ask a deeper question about what is the actual standard that OpenSSL complies with? Rich Salz has weighed in and claimed that C89 is the code base standard to adher to. It may also be worth at least talking about C99 compliance and addressing the poorly specified hint called "inline". We need to be aware that "inline" is just a gentle suggestion. It is a hint. There is no firm specification anywhere that tells any compiler ever anything about what to do with an "inline" function.  The problem here is not the use of "inline" but the thinking that OpenSSL should ever even try to use such a poorly specified thing.

**Comment by levitte** [comment]
> Please don't post 20,000 lines of repetition

B-b-b-but....  it was just a mere 1000 lines?
....
yeahok, you're right.  I've shortened down that comment substantially.

Regarding C standards and inline, let's also recognise that we're combining `static` and `inline`, makes it easier.  It was a mistake, though, to have static inline function definitions in the header files.

**Comment by rammishr** [comment]
Hi,
Sorry if this should have been posted separately. 

When i build openssl1.1.1b on solaris , it is successful. But while verifying the version " openssl version" it fails with below error. 
ld.so.1: openssl: fatal: relocation error: file /vobs_tools/prgs/src/openssl/solaris64/bin/openssl: symbol OPENSSL_sk_new_null: referenced symbol not found
Killed

The same machine works fine for openssl1.1.1. 
So, in 1.1.1b I can observe that OPENSSL_sk_new_null has been defined as below in safestack.h.
             pragma weak OPENSSL_sk_new_null
Can this be related ? Am I missing anything while Configure ?

**Comment by dtopham75** [comment]
I'm seeing the same behaviour as @rammishr with 1.1.1b, specifically when OpenSSL is configured for a static build (i.e. no-shared is passed to Configure). Running _openssl version_ fails with a similar error:

```
$ ./openssl version
ld.so.1: openssl: fatal: relocation error: file openssl: symbol OPENSSL_sk_num: referenced symbol not found
Killed
```
Not sure why it;s failing with a different symbol, maybe down to compiler version. 

This doesn't happen with a build of 1.1.1b configured to create dynamic libraries.

**Comment by mendobar** [comment]
I'm also seeing this issue when using the Sun Studio C 12.4 compiler on a Solaris 64-bit system running SunOS 5.10 and using openssl-1.1.1.c.  I need a static build.

When I run openssl, I get:
$ openssl
ld.so.1: openssl: fatal: relocation error: file openssl: symbol OPENSSL_sk_new_null: referenced symbol not found
Killed
$ 

I would prefer not to have to change the source code, so that the issue doesn't return in the next version that we fetch and untar.  

I see the bug "Error building non-portable against OpenSSL 1.1 #35807" above.  I did not find any mention of Solaris in that bug.  I'll try to resolve this, but I hope I don't have to change the code.

**Comment by blastwave** [comment]
OKay perhaps this is somehow a problem specific to x86_64 architecture but I don't see how. 
I see no such problems on Solaris 10 sparc : 

    beta$ ls -lapbE /usr/local/bin/openssl
    -rwxr-xr-x   1 root     root     26531808 2019-05-28 22:52:04.107498700 +0000 /usr/local/bin/openssl
    beta$ /usr/local/bin/openssl version
    OpenSSL 1.1.1c  28 May 2019
    beta$ 
    beta$ uname -a 
    SunOS beta 5.10 Generic_150400-65 sun4u sparc SUNW,SPARC-Enterprise
    beta$ 
    beta$ mcs -p /usr/local/bin/openssl | sort | tail -3
    acomp: Studio 12.6 Sun C 5.15 SunOS_sparc 2017/05/30
    cg: Studio 12.6 Compiler Common 12.6 SunOS_sparc 2017-05-30
    ld: Software Generation Utilities - Solaris Link Editors: 5.10-1.1522
    beta$ 
    
So build with Oracle Studio 12.6 all works great. One needs to glance at bugid 6912 ( whose silly idea was it to use github?? ) but other than that all is great. 

Have you tried with gcc 8.x ??

**Comment by lengauer** [comment]
I encounter the same issue as @mendobar when trying to build a static/no-shared build of OpenSSL 1.1.1c on Solaris10 sparc. I read through the several issues mentioned around here as well as some old mail chains and I am wondering if there is a sort of fix or workaround for this that has not been mentioned here? Because so far all the potential fixes or ideas I found did not solve the issue for me. E.g. removing the pragmas does not solve the issue was it just leads to the linker complaining about the missing symbols. Thus, I would appreciate any input or advice.

The only thing that works for me is to use "LD_PRELOAD=<path-to-libcrypto.so>" together with the openssl binary but I guess this is not really a nice solution. This would not really differ that much from a "shared" build... I achieved building the shared libraries such as "libcrypto.so" and "libssl.so" for my static build by doing a successive build with "shared" option so they yield from the exact same source.

Example of a functional call:
`LD_PRELOAD=<path-to-openssl-libs>/libcrypto.so openssl`

**Comment by lengauer** [comment]
It has been some time since my last comment and I figured it might be worth to share my "solution".
In order to make a static build of OpenSSL run on Solaris10 Sparc and not run into the error
```
ld.so.1: openssl: fatal: relocation error: file openssl: symbol OPENSSL_sk_new_null: referenced symbol not found
Killed
```
I found to workarounds:

1. Remove all the weak pragmas in "lhash.h" and "safestack.h" that are turned on if "__SUNPRO_C" is defined. The ensuing issue is that building the OpenSSL tests will fail. To avoid this I explicitly linked "libcrypto.a" for the cases that would not be able to run without it. Thus, in "openssl/test/build.info" is added dependencies using "DEPEND":
```
SOURCE[rsa_complex]=rsa_complex.c
INCLUDE[rsa_complex]=../include
DEPEND[rsa_complex]=../libcrypto.a
```
as well as:
```
IF[{- !$disabled{shared} -}]
   PROGRAMS_NO_INST=shlibloadtest
   SOURCE[shlibloadtest]=shlibloadtest.c
   INCLUDE[shlibloadtest]=../include ../crypto/include
   DEPEND[shlibloadtest]=../libcrypto.a
ENDIF
```
This way I was able to get the static OpenSSL build working on Solaris10 Sparc and my tests with the openssl binary were successful. 

2. **NOT recommended:** Each usage of the openssl binary has to be coupled with a "LD_PRELOAD=<path-to-openssl-libs>/libcrypto.so" to prevent the error shown above. I want to discourage the usage of this as it seems pretty tedious and might cause other issues...

Please let me know if you have any concerns, I just wanted to share what worked for me. As far as I can tell linking the static library should not impact the tests and I preferred to have a working openssl binary.

**Comment by TaAL7** [comment]
I ran into this issue on Solaris 11 and can confirm that the weak pragmas break the static lib. Simply removing them from "lhash.h" and "safestack.h" did the trick. After that both static and dynamic libs worked fine. So ironically the Solaris fixes break the Solaris build. Tested with Solaris Studio 12.4/5/6.

And on a side note, I second what some people said that OpenSSL should have its own bug system. After all it's one of the most important open source projects on the planet and not some hobby venture.
